src/hooks/ticker.ts

```diff
@@ -0,0 +1,164 @@
+import { useCallback, useEffect, useRef, useState } from 'react'
+
+// Only support specific token pairs until we find a method to validate token pairs
+export enum TokenPair {
+  BNBUSDT = 'bnbusdt',
+}
+
+/**
+ * @see https://binance-docs.github.io/apidocs/spot/en/#individual-symbol-ticker-streams
+ */
+export interface StreamData {
+  e: string
+  E: number
+  s: string
+  p: string
+  P: string
+  w: string
+  x: string
+  c: string
+  Q: string
+  b: string
+  B: string
+  a: string
+  A: string
+  o: string
+  h: string
+  l: string
+  v: string
+  q: string
+  O: number
+  C: number
+  F: number
+  L: number
+  n: number
+}
+
+export interface TickerStream {
+  eventType: string
+  eventTime: number
+  symbol: string
+  priceChange: number
+  priceChangePercent: number
+  weightAveragePrice: number
+  firstTrade: number
+  lastPrice: number
+  lastQuantity: number
+  bestBidPrice: number
+  bestBidQuantity: number
+  bestAskPrice: number
+  bestAskQuantity: number
+  openPrice: number
+  highPrice: number
+  lowPrice: number
+  totalTradedBaseAssetVolume: number
+  totalTradedQuoteAssetVolume: number
+  statisticsOpenTime: number
+  statisticsCloseTime: number
+  firstTradeId: number
+  lastTradeId: number
+  totalNumberOfTrades: number
+}
+
+export const useTokenPairTicker = (tokenPair: TokenPair, connectOnMount: boolean, retryCount = 3) => {
+  const retriesRemaining = useRef(retryCount)
+
+  // Chrome always returns an event code of 1006 so we have to keep track whether we closed the
+  // connection manually or not (1000)
+  const isManualClose = useRef(false)
+
+  // Use a ref instead of state so we mark the connection as closed immediately without
+  // triggering a re-render
+  const isConnected = useRef(false)
+  const [stream, setStream] = useState<TickerStream>(null)
+  const websocket = useRef<WebSocket>(null)
+
+  const connect = useCallback(() => {
+    if (navigator.onLine && !isConnected.current) {
+      const ws = new WebSocket(`wss://stream.binance.com:9443/ws/streams/${tokenPair}@ticker`)
+
+      ws.onopen = () => {
+        isConnected.current = true
+      }
+
+      ws.onclose = () => {
+        isConnected.current = false
+
+        if (retriesRemaining.current > 0 && !isManualClose.current && !isConnected.current) {
+          setTimeout(() => {
+            connect()
+            retriesRemaining.current -= 1
```

Comment:
I might be wrong, but isn't it the case that `retriesRemaining.current` will never reach 0?
- It starts with 3
- calls `connect()` here
- on line 135 it resets it back to 3 (`retriesRemaining.current = retryCount`) regardless if it re-connected or not
- decrements it by 1 here which makes it 2
- next time it retires it sets it back to 3 and so on

Reply:
Also another less likely scenario that came into my head:
- It tries to reconnect
- Creates new WebSocket instance
- It opens connection and then quickly closes it within 5 seconds setTimeout
- Original retry setTimeout will see isConnected.current as false and will try to reconnect again
- The second instances that opened and closed fires its own setTimeout and tries to reconnect in parallel with 1st reconnect routine.

Not sure how likely it is and if its worth researching, but just a though.